### PR TITLE
Support migration from ChirpStack V4

### LINKED ISSUE #58: Support migration from ChirpStack V4
#### Summary

Support exporting devices from https://www.chirpstack.io/docs/

#### Why do we need this?

New version of ChirpStack was released. [v4 breaking changes](https://www.chirpstack.io/docs/v4-breaking-changes.html)

#### What is already there? What do you see now?

We support migration from ChirpStack V3.

#### What is missing? What do you want to see?

Support for device migration from ChirpStack V4.

#### Environment

[v0.9.0-rc0](https://github.com/TheThingsNetwork/lorawan-stack-migrate/releases/tag/v0.9.0-rc0)

#### How do you propose to implement this?

Use gRPC API or REST API to export devices.

#### How do you propose to test this?

Test locally with [chirpstack/chirpstack-docker](https://github.com/chirpstack/chirpstack-docker) which works out of the box now (If running V3 needs adding execute permissions to database init files since it was fixed after switching to V4).

#### Can you do this yourself and submit a Pull Request?

I can do this myself.

### PR DESCRIPTION
**Pull request by happyRip**
#### Summary

Closes #58

#### Changes

- Use ChirpStack V4 API
- Unify flag names between sources

#### Testing

Tested with local ChirpStack v4 instance.

##### Regressions

Flag name changes affect all sources. TTS and ChirpStack sources were tested.

#### Notes for Reviewers

ChirpStack v3 and v4 protobuf APIs conflict with each other and cannot be compiled into the same binary.

See https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/75/files#r1232266208 since I'm not sure if this is handled correctly

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

File: pkg/source/chirpstack/source.go [hunk 1]
```diff
@@ -205,30 +192,23 @@ func (p *Source) ExportDevice(devEui string) (*ttnpb.EndDevice, error) {
...
 			dev.MacSettings.Rx1Delay = &ttnpb.RxDelayValue{
-				Value: ttnpb.RxDelay(devProfile.RxDelay_1),
+				Value: ttnpb.RxDelay(delay),
 			}
 		}
-		if devProfile.RxDrOffset_1 >= 0 {
-			dev.MacSettings.DesiredRx1DataRateOffset = &ttnpb.DataRateOffsetValue{
-				Value: ttnpb.DataRateOffset(devProfile.RxDrOffset_1),
-			}
+		dev.MacSettings.DesiredRx1DataRateOffset = &ttnpb.DataRateOffsetValue{
```
Comment (adriansmares) [inline pkg/source/chirpstack/source.go:207]: I would set this only if it is greater than zero as it was originally done. Also are we sure it is desired ?

**Review by adriansmares — COMMENTED** [review]
The changes to the environment variables and flag names should also be updated in the readme.

---

For the context on why we cannot have two sources: both the v3 and the v4 API use the same protobuf `package` statements and message names, but with different contents. This is not supported by the protobuf library, as the message definition registry is global, and ChirpStack messages have same path for two different messages

My proposal here was that since the v3 ChirpStack won't receive any further development, that we can just refer people to older versions of the migration tool. Another solution would be to have a build tag dedicated just to the v3 API, but I don't know if that is worth the effort and the complexity.

File: pkg/source/chirpstack/util.go [hunk 2]
```diff
@@ -130,7 +114,7 @@ func (p *Source) getApplicationIDByName(name string) (int64, error) {
 		}
 		for _, appListItem := range resp.Result {
 			if appListItem.Name == name {
-				return appListItem.Id, nil
+				return strconv.ParseInt(appListItem.Id, 10, 64)
```
[the remark this answers is not in the document]
Comment (happyRip) [reply pkg/source/chirpstack/util.go]: Left the temporary fix there and missed it when looking through the changes.

CSv4 IDs are UUID now, so i ditched the `strconv.FormatInt` logic.

File: pkg/source/chirpstack/source.go (hunk 1 above)
Comment (happyRip) [reply pkg/source/chirpstack/source.go:207]: `DesiredRx1DataRateOffset ` is unsigned now, so I removed the `>=` condition. Should those be kept for the sake of verbosity?

> is it a desired parameter

It's hard for me to judge that, I didn't want to remove the features we had here previously and compared the values from CS v3 to v4 migration script.

File: pkg/source/chirpstack/source.go [hunk 3]
```diff
@@ -160,39 +149,37 @@ func (p *Source) ExportDevice(devEui string) (*ttnpb.EndDevice, error) {
...
+		dev.MacSettings.StatusTimePeriodicity = durationpb.New(time.Duration(d))
 	}
 
 	// Frequency Plan
 	dev.FrequencyPlanId = p.FrequencyPlanID
 
 	// General
-	switch devProfile.MacVersion {
-	case "1.0.0":
+	switch devProfile.MacVersion.String() {
```
Comment (adriansmares) [inline pkg/source/chirpstack/source.go:160]: Let's use the enum instead of comparing strings.

File: pkg/source/chirpstack/source.go [hunk 4]
```diff
@@ -160,39 +149,37 @@ func (p *Source) ExportDevice(devEui string) (*ttnpb.EndDevice, error) {
...
 		dev.LorawanPhyVersion = ttnpb.PHYVersion_TS001_V1_0
-	case "1.0.1":
+	case "LORAWAN_1_0_1":
 		dev.LorawanVersion = ttnpb.MACVersion_MAC_V1_0_1
 		dev.LorawanPhyVersion = ttnpb.PHYVersion_TS001_V1_0_1
-	case "1.0.2":
+	case "LORAWAN_1_0_2":
 		dev.LorawanVersion = ttnpb.MACVersion_MAC_V1_0_2
-		switch devProfile.RegParamsRevision {
+		switch devProfile.RegParamsRevision.String() {
```
Comment (adriansmares) [inline pkg/source/chirpstack/source.go:169]: Let's use the enum instead of comparing strings.

File: pkg/source/chirpstack/source.go (hunk 1 above)
Comment (adriansmares) [reply pkg/source/chirpstack/source.go:207]: I would set this only if it is greater than 0, and I believe that the original implementation was wrong, and this is actually `Rx1DataRateOffset`, not `DesiredRx1DataRateOffset`.

File: pkg/source/chirpstack/source.go [hunk 5]
```diff
@@ -205,30 +192,23 @@ func (p *Source) ExportDevice(devEui string) (*ttnpb.EndDevice, error) {
...
+			Value: ttnpb.DataRateOffset(devProfile.AbpRx1DrOffset),
 		}
-		if devProfile.RxDatarate_2 >= 0 {
-			dev.MacSettings.DesiredRx2DataRateIndex = &ttnpb.DataRateIndexValue{
-				Value: ttnpb.DataRateIndex(devProfile.RxDatarate_2),
-			}
-		}
-		for _, freq := range devProfile.FactoryPresetFreqs {
-			dev.MacSettings.FactoryPresetFrequencies = append(dev.MacSettings.FactoryPresetFrequencies, uint64(freq))
+		dev.MacSettings.DesiredRx2DataRateIndex = &ttnpb.DataRateIndexValue{
```
[the remark this answers is not in the document]
Comment (adriansmares) [reply pkg/source/chirpstack/source.go:210]: This seems to be `Rx2DataRateIndex`, not `DesiredRx2DataRateIndex`

File: pkg/source/chirpstack/source.go [hunk 6]
```diff
@@ -289,30 +266,28 @@ func (p *Source) ExportDevice(devEui string) (*ttnpb.EndDevice, error) {
 	}
 
 	// Payload formatters
-	switch devProfile.PayloadCodec {
+	switch devProfile.PayloadCodecRuntime.String() {
```
Comment (adriansmares) [inline pkg/source/chirpstack/source.go:269]: Let's use their enum.

**Review by adriansmares — APPROVED** [review]
LGTM. @johanstokking do you have any objections to dropping support completely for ChirpStack v3 going forward (see the discussion above) ?

The only alternative that I could think of was to basically maintain our own generated code version of their proto files, with a `package` statement added manually in order to avoid message collisions. We would then have to edit the generated gRPC code in order to edit out the package from the RPC paths. It is very, very messy, but doable.

**Comment by johanstokking** [comment]
I think this is acceptable going forward indeed. I don't see another clean solution either.

**Comment by KrishnaIyer** [comment]
@happyRip: What's the status of addressing the review comments?

**Comment by KrishnaIyer** [comment]
replaced by https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/112

**Pull request closed without merging**